src/connector/src/sink/dynamodb.rs

```diff
@@ -0,0 +1,298 @@
+            Ok(())
+        }
+    }
+}
+
+pub struct DynamoDbSinkWriter {
+    payload_writer: DynamoDbPayloadWriter,
+    formatter: DynamoDbFormatter,
+}
+
+impl DynamoDbSinkWriter {
+    pub async fn new(config: DynamoDbConfig, schema: Schema) -> Result<Self> {
+        let client = config.common.build_client().await?;
+
+        let payload_writer = DynamoDbPayloadWriter {
+            insert_items: HashMap::new(),
+            delete_items: HashMap::new(),
+            client,
+            table: config.common.table,
+        };
+
+        Ok(Self {
+            payload_writer,
+            formatter: DynamoDbFormatter { schema },
+        })
+    }
+
+    async fn write_chunk_inner(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            match op {
+                Op::Insert | Op::UpdateInsert => {
+                    for (k, v) in self.formatter.format_row(row) {
+                        self.payload_writer.write_one_insert(k, v);
+                    }
+                }
+                Op::Delete | Op::UpdateDelete => {
+                    for (k, v) in self.formatter.format_row(row) {
+                        self.payload_writer.write_one_delete(k, v);
+                    }
+                }
+            }
+        }
+        self.payload_writer.write_chunk().await
+    }
+}
+
+impl AsyncTruncateSinkWriter for DynamoDbSinkWriter {
+    async fn write_chunk<'a>(
+        &'a mut self,
+        chunk: StreamChunk,
+        _add_future: DeliveryFutureManagerAddFuture<'a, Self::DeliveryFuture>,
+    ) -> Result<()> {
+        self.write_chunk_inner(chunk).await
+    }
+}
+
+struct DynamoDbFormatter {
+    schema: Schema,
+}
+
+impl DynamoDbFormatter {
+    fn format_row(&self, row: RowRef<'_>) -> Vec<(String, AttributeValue)> {
+        let mut items = Vec::new();
+        for (scalar, field) in row.iter().zip_eq_debug((self.schema.clone()).into_fields()) {
+            let attr_value = map_data_type(scalar);
+            items.push((field.name, attr_value));
+        }
+
+        items
+    }
+}
+
+fn map_data_type(scalar: Option<ScalarRefImpl<'_>>) -> AttributeValue {
+    match scalar {
+        None => AttributeValue::Null(true),
+        Some(s) => match s {
+            number @ (ScalarRefImpl::Int16(_)
+            | ScalarRefImpl::Int32(_)
+            | ScalarRefImpl::Int64(_)
+            | ScalarRefImpl::Int256(_)
+            | ScalarRefImpl::Float32(_)
+            | ScalarRefImpl::Float64(_)
+            | ScalarRefImpl::Decimal(_)
+            | ScalarRefImpl::Serial(_)) => AttributeValue::N(number.to_text()),
+            string @ (ScalarRefImpl::Utf8(_)
+            | ScalarRefImpl::Interval(_)
+            | ScalarRefImpl::Date(_)
+            | ScalarRefImpl::Time(_)
+            | ScalarRefImpl::Timestamp(_)
+            | ScalarRefImpl::Timestamptz(_)
+            | ScalarRefImpl::Struct(_)
+            | ScalarRefImpl::Jsonb(_)) => AttributeValue::S(string.to_text()),
```

Comment:
* `struct` may be better mapped to DynamoDB [Map](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html#HowItWorks.DataTypes.Document.Map)
* `jsonb` can be mapped as `string` for now, and latter extended to a **dynamic** recursive one similar to [#11699](https://github.com/risingwavelabs/risingwave/issues/11699). It is your call whether the latter dynamic one shall be the default here instead. Just bringing the alternative option to attention.

Reply:
- I'm not sure, the struct here must be named, since in `create sink`?
- Does this mean providing an option for `jsonb` format?

Reply:
You can consider struct as a list of named key values pairs, right?

```
struct <
  id varchar,
  name varchar
>
```

```
Map<String, Value> {
  "id": "123",
  "name": "jinser"
}
```

Reply:
Yes, I think the current implementation is like this